pkg/operator/startupmonitorreadiness/readiness_checks.go

```diff
@@ -0,0 +1,259 @@
+}
+
+// doKubeHealthCheck performs an HTTP check against healthz?verbose=true endpoint
+//  returns true, "", "", on HTTP 200
+//  returns false, "Unhealthy", EntireResponseBody (if any) on HTTP != 200
+//  returns false, "UnhealthyError", EntireResponseBody (if any) in case of any error or timeout
+func doKubeHealthCheck(ctx context.Context, client *http.Client, rawURL string) (bool, string, string) {
+	return doHTTPCheckAndTransform(ctx, client, fmt.Sprintf("%s/healthz?verbose=true", rawURL), "Unhealthy", doHTTPCheck)
+}
+
+// doETCDHealthCheck performs an HTTP check against healthz/etcd endpoint
+//  returns true, "", "", on HTTP 200
+//  returns false, "EtcdUnhealthy", EntireResponseBody (if any) on HTTP != 200
+//  returns false, "EtcdUnhealthyError", EntireResponseBody (if any) in case of any error or timeout
+func doETCDHealthCheck(ctx context.Context, client *http.Client, rawURL string) (bool, string, string) {
+	return doHTTPCheckAndTransform(ctx, client, fmt.Sprintf("%s/healthz/etcd", rawURL), "EtcdUnhealthy", doHTTPCheck)
+}
+
+func doHTTPCheckAndTransform(ctx context.Context, client *http.Client, rawURL string, checkName string, httpCheckFn func(ctx context.Context, client *http.Client, rawURL string) (int, string, error)) (bool, string, string) {
+	statusCode, response, err := httpCheckFn(ctx, client, rawURL)
+	if err != nil {
+		errMsg := fmt.Sprintf("falied while performing the check due to %v", err)
+		if len(response) > 0 {
+			errMsg = fmt.Sprintf("%v, a response from the server was %v", errMsg, response)
+		}
+		return false, fmt.Sprintf("%vError", checkName), errMsg
+	}
+	if statusCode != http.StatusOK {
+		return false, checkName, response
+	}
+
+	return true, "", ""
+}
+
+func doHTTPCheck(ctx context.Context, client *http.Client, rawURL string) (int, string, error) {
+	targetURL, err := url.Parse(rawURL)
+	if err != nil {
+		return 0, "", err
+	}
+	newReq, err := http.NewRequestWithContext(ctx, "GET", targetURL.String(), nil)
+	if err != nil {
+		return 0, "", err
+	}
+
+	resp, err := client.Do(newReq)
+	if err != nil {
+		return 0, "", err
+	}
+	defer resp.Body.Close()
+
+	// we expect small responses from the server
+	// so it is okay to read the entire body
+	rawResponse, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return 0, "", fmt.Errorf("error while reading body from %v, err %v", targetURL.String(), err)
+	}
+
+	return resp.StatusCode, string(rawResponse), nil
+}
+
+func createHTTPClient(responseTimeout time.Duration, restConfig *rest.Config) (*http.Client, error) {
+	transportConfig, err := restConfig.TransportConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	tlsConfig, err := transport.TLSConfigFor(transportConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &http.Client{
+		Transport: utilnet.SetTransportDefaults(&http.Transport{
+			TLSClientConfig: tlsConfig,
+		}),
+		Timeout: responseTimeout,
```

Comment:
http.Client.Timeout may timeout even if the connection was sucesfull

https://cs.opensource.google/go/go/+/refs/tags/go1.16.6:src/net/http/client.go;l=90-94

I fixed recently a bug in kops because of this, https://github.com/kubernetes/kops/pull/11884/files#diff-8a8f23d88ae96f6029ac0558e279ed8850568082c913852bfd45c2b54e26166fR81-R93 

We are operating in a 2 seconds timeframe that seems to short, but still can race.
What about having more granularity (I'm making up the values here)?
```
	httpClient := &http.Client{
		Transport: &http.Transport{
			DialContext: (&net.Dialer{
				Timeout:   5 * time.Second,
				KeepAlive: 5 * time.Second,
			}).DialContext,
			TLSHandshakeTimeout:   5 * time.Second,
			ResponseHeaderTimeout: 5 * time.Second,
			IdleConnTimeout:       5 * time.Second,
		},
```

Reply:
This http client will be talking to Kube API over localhost. In case of an error/timeout requests will be retired for 5 min. 

Do we still think 2 seconds is too short?

Reply:
ah, ok, this is always localhost, maybe I'm being  too paranoid :smile: 
these requests doesn't go through the apf thing, right? there is no chance the server can take more than 2 seconds to reply? The different timeouts help to known which part timed out exactly, but as you say this may not apply here

Reply:
> these requests doesn't go through the apf thing, right?

it does, we will authN as "system:master" which always is privileged.

> there is no chance the server can take more than 2 seconds to reply?

there is, for example getting a pod might take much longer if etcd is slow for example. I can increase the timeout to 5 seconds. 

The purpose of these checks is to assess the readiness of the Kube API server and fallback to the previous version in case of any issues. If the API server is unable to serve requests because the connection to etcd is "slow" we should fail.

Reply:
I saw the Pod.List() and other api calls take 2 seconds on busy CIs, can this be a concern here?

Reply:
ahh, this uses the kubeclient not the http.client